crates/sui-json-rpc/src/coin_api.rs

```diff
@@ -288,41 +209,706 @@ impl CoinReadApiServer for CoinReadApi {
     #[instrument(skip(self))]
     async fn get_coin_metadata(&self, coin_type: String) -> RpcResult<Option<SuiCoinMetadata>> {
         with_tracing!(async move {
-            let coin_struct = parse_sui_struct_tag(&coin_type)?;
-
-            let metadata_object = self
-                .find_package_object(
-                    &coin_struct.address.into(),
-                    CoinMetadata::type_(coin_struct),
-                )
-                .await
-                .ok();
-
-            Ok(metadata_object.and_then(|v: Object| v.try_into().ok()))
+            let coin_struct = parse_sui_struct_tag(&coin_type).map_err(|_| {
+                Error::SuiRpcInputError(SuiRpcInputError::CannotParseSuiStructTag(coin_type))
+            })?;
+            Ok(self.internal.get_coin_metadata(coin_struct).await?)
         })
     }
 
     #[instrument(skip(self))]
     async fn get_total_supply(&self, coin_type: String) -> RpcResult<Supply> {
         with_tracing!(async move {
-            let coin_struct = parse_sui_struct_tag(&coin_type)?;
+            let coin_struct = parse_sui_struct_tag(&coin_type).map_err(|_| {
+                Error::SuiRpcInputError(SuiRpcInputError::CannotParseSuiStructTag(coin_type))
+            })?;
 
             Ok(if GAS::is_gas(&coin_struct) {
                 Supply { value: 0 }
             } else {
-                let treasury_cap_object = self
-                    .find_package_object(
-                        &coin_struct.address.into(),
-                        TreasuryCap::type_(coin_struct),
-                    )
-                    .await?;
-
-                let treasury_cap = TreasuryCap::from_bcs_bytes(
-                    treasury_cap_object.data.try_as_move().unwrap().contents(),
-                )
-                .map_err(Error::from)?;
+                let treasury_cap = self.internal.get_treasury_cap(coin_struct).await?;
                 treasury_cap.total_supply
             })
         })
     }
 }
+
+#[cfg_attr(test, automock)]
+#[async_trait]
+pub trait StateServiceWrapper {
```

Comment:
I'm struggling to understand what `Service` and `Wrapper` mean in this name.  Let's name this trait for what it is used for (e.g. what consumes this trait, and why?)

Reply:
Also, we should document what these traits abstract away (just a couple of lines of doc comment is fine).